### PR TITLE
[Feature] cp4data: Allow install of CP4D v3.5

### PR DESCRIPTION
**Pull request by johandry**
This PR uses the latest CP4Data module which allow us to install version 3.5.
The code add the required variables for v3.5 in the terraform code, documentation and schematics json template.

### CODE REVIEW

**Comment by johandry** [comment]
The installation is facing with the following errors:

```
[ERROR] [2021-01-15 03:35:11-0492] Error trying to reuse blob sha256:ec1681b6a383e4ecedbeddd5abc596f3de835aed6db39a735f62395c8edbff30 at destination: error pinging docker registry image-registry-openshift-image-registry.cpd35-sandbox-cluster-c0b572361ba41c9eef42d4d51297b04b-0000.us-south.containers.appdomain.cloud: Get "http://image-registry-openshift-image-registry.cpd35-sandbox-cluster-c0b572361ba41c9eef42d4d51297b04b-0000.us-south.containers.appdomain.cloud/v2/": dial tcp: lookup image-registry-openshift-image-registry.cpd35-sandbox-cluster-c0b572361ba41c9eef42d4d51297b04b-0000.us-south.containers.appdomain.cloud on 172.21.0.10:53: no such host
```

```
[ERROR] [2021-01-15 03:35:12-0094] Exiting due to error (Error in handling image request regarding privatecloud-usermgmt:3.5.1-x86_64-352 from registry docker://cp.icr.io/cp/cpd/privatecloud-usermgmt:3.5.1-x86_64-352 after 4 retries. Abort). Please check /cpd-cli-workspace/logs/CPD-2021-01-15T03-34-32.log for details
[ERROR] 2021-01-15T03:35:12.097418Z Execution error:  exit status 1
```

**Comment by johandry** [comment]
Errors removed, installation ran successfully 😄 

```
 [INFO] [2021-01-15 20:33:41-0904] *** Install/Upgrade for assembly lite completed successfully ***

 Installation of assembly lite is successfully completed
 Unable to use a TTY - input is not a terminal or the right kind of file
  The Services installed successfully from the chosen services : + +lite
  The Services failed to install from the chosen services: None
 Unable to use a TTY - input is not a terminal or the right kind of file
 tar: Removing leading `/' from member names
 route.route.openshift.io "cloudpak4data-cpd" deleted
 route.route.openshift.io/cloudpak4data-cpd created
 ##################################################################################
 Installation finished
 Access Cloud Pak for Data console using the address:
 cloudpak4data-cpd-cloudpak4data.cpd35-sandbox-cluster-c0b572361ba41c9eef42d4d51297b04b-0000.us-south.containers.appdomain.cloud
 ##################################################################################
module.cp4data.null_resource.install_cp4data[0]: Creation complete after 22m47s [id=2860690782089783056]
```

**Comment by johandry** [comment]
The test of this code was done locally using Make with Terraform.
The Schematics is being tested right now, so the creation of the Workspace works.

**Comment by cambelem** [comment]
The first error I've come across using the make install doc is clarifying the terraform versioning. I noticed that In the doc, it doesn't state the best version of terraform that user should use. I'm currently using terraform 14 and it stops with the following error below:

```
Initializing provider plugins...
- Finding latest version of hashicorp/null...
- Finding latest version of hashicorp/ibm...
- Finding latest version of hashicorp/external...
- Finding latest version of hashicorp/local...
- Installing hashicorp/null v3.0.0...
- Installed hashicorp/null v3.0.0 (signed by HashiCorp)
- Installing hashicorp/external v2.0.0...
- Installed hashicorp/external v2.0.0 (signed by HashiCorp)
- Installing hashicorp/local v2.0.0...
- Installed hashicorp/local v2.0.0 (signed by HashiCorp)

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider hashicorp/ibm:
provider registry registry.terraform.io does not have a provider named
registry.terraform.io/hashicorp/ibm
```

I've swapped to Terraform v12 and it is currently running. I'll update this comment once it is complete or if there are any more errors.

File: terraform/Using_Make.md [hunk 1]
```diff
@@ -130,29 +139,31 @@ Make sure everything that provisioning the Cloud Pak using Schematics works befo
...
 
-4. When ready click on "**Generate plan**" or execute: `make plan-sch`, you will get the link to the plan output/log
+3. When ready click on "**Generate plan**" or execute: `make plan-sch`, you will get the link to the plan output/log
 
    ```bash
    make plan-sch
    ```
 
-5. When the plan successfully finish, click on "**Apply plan**" or execute: `make apply-sch`, you will get the link to the plan output/log.
+4. When the plan successfully finish, click on "**Apply plan**" or execute: `make apply-sch`, you will get the link to the plan output/log.
```
Comment (annumberhocker) [inline terraform/Using_Make.md:166]: This needs to say "When the plan successfully finishes, ..."

File: terraform/cp4data/terraform.tfvars [hunk 2]
```diff
@@ -7,14 +7,16 @@
...
-environment  = "sandbox"
-region       = "us-south"
-// Using Humio account by default.  This resource group exists only on this account
+project_name   = "cloud-pack"
+owner          = "anonymous"
+environment    = "sandbox"
+region         = "us-south"
 resource_group = "cloud-pak-sandbox"
-// resource_group = "cloud-pak-sandbox"
+// resource_group = "cloud-pak-sandbox-ibm"
```
Comment (annumberhocker) [inline terraform/cp4data/terraform.tfvars:19]: Add these lines above 
`resource_group = "cloud-pak-sandbox"` like in `cp4app/terraform.tfvars`.  It makes things clearer.

```
// If using Humio account:
// resource_group = "cloud-pak-sandbox-ibm"
// If using standard partner account:
```

**Review by annumberhocker — CHANGES_REQUESTED** [review]

**Comment by cambelem** [comment]
I came across this error after the cluster was created.

`[INSTALLER : INFO ] Installing Cloud Pak for Data
[INFO] [2021-01-15 16:55:52-0377] --download-path not specified, creating workspace cpd-workspace
[INFO] [2021-01-15 16:55:52-0400] Enter credentials for default-route-openshift-image-registry.emc-cp-data-sandbox-clu-c0b572361ba41c9eef42d4d51297b04b-0000.us-south.containers.appdomain.cloud/cloudpak4data

[ERROR] [2021-01-15 16:55:56-0803] Invalid storage class 'ibmc-file-custom-gold-gid'. Exiting
[INSTALLER : ERROR] failed to install Cloud Pak for Data`

I have deleted and cleaned everything and will be running this again to ensure there wasn't a problem on my end.

**Comment by johandry** [comment]
The PR was reviewed and approved by Shivam Solanki. It was also reviewed by Eric Cambel with a minor suggestion. I also found some improvements to be done in a future PR.

Suggestions and improvements:
- Add to the documentation that Terraform 0.12 is the required version - by Eric
- Take the URL entry point of cp4data and report it as output variable - by Johandry
- Report the default credentials `admin`, `password` as output variables - by Johandry
- Wait for the routes to be ready before install cp4data. This may be the cause of the failure of the first execution on Schematics - by Johandry
- The Storage Class for 3.0 and 3.5 are different, make it clear in documentation or code - by Johandry

Advance changes, an entire PR is required for each issue:
- Support other storages for 3.5
- Migrate cpd 3.0 to the same way 3.5 is installed (if 3.0 will be still in use)
- Make the input variables simpler. Some input variables for 3.0 and 3.5 are not common, this is confusing to the user. A possible solution is to make a module for 3.5 and another for 3.0

**Pull request merged**